### PR TITLE
Adds genesis block program, in-memory ledger to `snarkvm-ledger`, and unifies `DPC`

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR introduces a basic in-memory ledger to allow us to create the genesis block.

- Adds genesis block program and script
- Adds an unoptimized in-memory ledger to `snarkvm-ledger`
- Unifies `DPC` structs and refactors parameters in a `parameters` folder
- Adds SNARK proving key and verifying key getter methods in `snarkvm_dpc::Parameters` trait
- Samples a genesis block for `testnet1` and `testnet2`

### CODE REVIEW

**Review by weikengchen — APPROVED** [review]
Looks reasonable.

File: dpc/src/parameters/testnet2.rs [hunk 1]
```diff
@@ -77,6 +77,41 @@ macro_rules! dpc_setup {
...
+                            $parameter::load_bytes().expect(&format!("Failed to load parameter bytes for {}", $message)).as_slice(),
+                        ).expect(&format!("Failed to read {} from bytes", $message)))
+                    })
+                }
+                false => &None,
+            }
+        }
+    };
+}
+
```
Comment (ljedrz) [inline dpc/src/parameters/testnet2.rs:114]: nit, can be done later: these macros can apply to both `testnet1` and `testnet2` with an extra added argument

File: dpc/src/program/noop_program.rs [hunk 2]
```diff
@@ -74,33 +71,10 @@ impl<C: Parameters> ProgramScheme for NoopProgram<C> {
...
-            Network::Testnet2 => <Self::ProofSystem as SNARK>::VerifyingKey::read_le(
-                snarkvm_parameters::testnet2::NoopProgramSNARKVKParameters::load_bytes()?.as_slice(),
-            )?,
-            _ => {
-                unimplemented!()
-            }
-        };
-
+        let proving_key = C::noop_program_proving_key().clone();
+        let verifying_key = C::noop_program_verifying_key().clone();
```
Comment (ljedrz) [inline dpc/src/program/noop_program.rs:77]: nit: `NoopProgram` can probably just contain those 2 `&'static`s without the need to ever clone them.

File: parameters/examples/genesis.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+pub fn generate<C: Parameters>(recipient: &Address<C>, value: u64) -> Result<(Vec<u8>, Vec<u8>), DPCError> {
+    let rng = &mut thread_rng();
+
+    // TODO (howardwu): Deprecate this in favor of a simple struct with 2 Merkle trees.
+    let temporary_ledger = {
+        let mut temporary_path = std::env::temp_dir();
+        temporary_path.push(format!("./test_db-{}", rand::thread_rng().gen::<usize>()));
+
+        Ledger::<C, MemDb>::new(Some(&temporary_path), Block {
```
Comment (ljedrz) [inline parameters/examples/genesis.rs:41]: the path should not be needed if `MemDb` is used, i.e. this call should rather be:
```
Ledger::<C, MemDb>::new(None, Block {
```

Comment (howardwu) [reply parameters/examples/genesis.rs:41]: Good catch, will fix

File: dpc/src/parameters/testnet2.rs (hunk 1 above)
Comment (howardwu) [reply dpc/src/parameters/testnet2.rs:114]: Discussed offline, attempted and unable to make it work for some odd reason. @ljedrz will investigate and we will follow up with a later PR if a resolution is found.

File: dpc/src/program/noop_program.rs (hunk 2 above)
Comment (howardwu) [reply dpc/src/program/noop_program.rs:77]: Discussed offline, Derivative doesn't cooperate well with this change.

**Review by ljedrz — APPROVED** [review]
LGTM :+1:

**Pull request merged**